### PR TITLE
KAFKA-4937: Batch offset fetches in the Consumer

### PR DESCRIPTION
**Pull request by dguy**
change `consumer.position` so that it always updates any partitions that need an update. Keep track of partitions that `seekToBeginning` in `StoreChangeLogReader` and do the `consumer.position` call after all `seekToBeginning` calls.

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang WDYT?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -371,15 +368,24 @@ else if (strategy == OffsetResetStrategy.LATEST)
...
+        for (final TopicPartition partition : partitions) {
+            offsetResets.put(partition, offsetResetStrategy(partition));
+        }
+        final Map<TopicPartition, OffsetData> offsetsByTimes = retrieveOffsetsByTimes(offsetResets, Long.MAX_VALUE, false);
+        for (final TopicPartition partition : partitions) {
+            final OffsetData offsetData = offsetsByTimes.get(partition);
+            if (offsetData == null) {
+                throw new NoOffsetForPartitionException(partition);
+            }
+            if (subscriptions.isAssigned(partition)) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Better keep the comment of `// we might lose the assignment while fetching the offset, so check it is still active`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -244,11 +247,11 @@ public void updateFetchPositions(Set<TopicPartition> partitions) {
                 continue;
 
             if (subscriptions.isOffsetResetNeeded(tp)) {
-                resetOffset(tp);
+                resetOffsets(Collections.singleton(tp));
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Currently `updateFetchPositions` is always called after `resetOffsetsIfNeeded` so these branch should not be hit in practice other than unit tests, but I think we can still batch them if necessary or simply throw an exception if `subscriptions.hasValidPosition(tp) && subscriptions.committed(tp) == null`. WDYT @hachikuji ?

**Review by guozhangwang — COMMENTED** [review]
LGTM overall. cc @hachikuji

**Comment by guozhangwang** [comment]
While reviewing the code I feel there are some code cleanup / refactoring that we can do in `Fetcher` (not in the scope of this PR). 

For example, `updateFetchPositions()` is only triggered in `KafkaConsumer#updateFetchPositions()`, in which `resetOffsetsIfNeeded()` is always triggered beforehand, so inside that function these two conditions `subscriptions.committed(tp) == null` and `subscriptions.isOffsetResetNeeded(tp)` should never be true unless it is used in unit tests; maybe we can tighten the logic a bit for such cases in a separate PR?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: The ordering is a little tricky when setting the offsets. The logic is like this:

1. Fetch offsets which have been explicitly requested by the user through `seekToBeginning` or `seekToEnd`. (We do this first to avoid unnecessary dependence on the group coordinator for consumers using manual assignment.)
2. Fetch committed offsets for any positions which haven't been updated.
3. Now reset offsets for any partitions which are left (i.e. those that were not explicitly seeked on and which have no committed offset).

So I think we still need to handle this here. My question for this patch is which case are you trying to optimize for? The case where you have done a `seek` on all the partitions?

**Comment by hachikuji** [comment]
One minor comment: can we change the title to emphasize that the batching is done in the consumer?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 2 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: That's right. Thanks for the explanation.

**Comment by dguy** [comment]
Updated. Thanks @hachikuji @guozhangwang

**Comment by guozhangwang** [comment]
The newly added unit test hangs (cc @bbejeck ):

```

org.apache.kafka.streams.integration.RegexSourceIntegrationTest > shouldAddStateStoreToRegexDefinedSource STARTED
Build timed out (after 180 minutes). Marking the build as aborted.
Build was aborted
Recording test results
Setting GRADLE_2_4_RC_2_HOME=/home/jenkins/jenkins-slave/tools/hudson.plugins.gradle.GradleInstallation/Gradle_2.4-rc-2
```

**Comment by guozhangwang** [comment]
retest this please.

**Comment by dguy** [comment]
@guozhangwang i think we have a general build hanging issue with some integration tests. But we should at least wait to see the builds go green

**Comment by bbejeck** [comment]
@guozhangwang <https://github.com/guozhangwang> if this continues to be a
problem can we disable this test? I won't be able to dig into it until later

EDIT: Running clean build locally trying to reproduce the issue

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
@ijuma can you have a look at this one too, please?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 3]
```diff
@@ -355,13 +358,7 @@ else if (error.exception() instanceof RetriableException)
...
-    /**
-     * Reset offsets for the given partition using the offset reset strategy.
-     *
-     * @param partition The given partition that needs reset offset
-     * @throws org.apache.kafka.clients.consumer.NoOffsetForPartitionException If no offset reset strategy is defined
-     */
-    private void resetOffset(TopicPartition partition) {
+    private long offsetResetStrategy(final TopicPartition partition) {
         OffsetResetStrategy strategy = subscriptions.resetStrategy(partition);
         log.debug("Resetting offset for partition {} to {} offset.", partition, strategy.name().toLowerCase(Locale.ROOT));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I think we should probably move this log statement since this method doesn't actually reset offsets.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -355,13 +358,7 @@ else if (error.exception() instanceof RetriableException)
...
     }
 
-    /**
-     * Reset offsets for the given partition using the offset reset strategy.
-     *
-     * @param partition The given partition that needs reset offset
-     * @throws org.apache.kafka.clients.consumer.NoOffsetForPartitionException If no offset reset strategy is defined
-     */
-    private void resetOffset(TopicPartition partition) {
+    private long offsetResetStrategy(final TopicPartition partition) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: `offsetResetStrategyTimestamp`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 5]
```diff
@@ -355,13 +358,7 @@ else if (error.exception() instanceof RetriableException)
             return client.send(node, request);
     }
 
-    /**
-     * Reset offsets for the given partition using the offset reset strategy.
-     *
-     * @param partition The given partition that needs reset offset
-     * @throws org.apache.kafka.clients.consumer.NoOffsetForPartitionException If no offset reset strategy is defined
-     */
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:363]: Shouldn't this comment be moved to `resetOffsets` (and updated)?

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, left some minor comments.

**Comment by dguy** [comment]
@ijuma hopefully this is good to go now. thanks

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 6]
```diff
@@ -1269,7 +1269,7 @@ public long position(TopicPartition partition) {
                 throw new IllegalArgumentException("You can only check the position for partitions assigned to this consumer.");
             Long offset = this.subscriptions.position(partition);
             if (offset == null) {
-                updateFetchPositions(Collections.singleton(partition));
+                updateFetchPositions(subscriptions.assignedPartitions());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: What is the motivation for this change? It seems a little odd that we update the fetch positions for all assigned partitions if the position for one partition is `null`.

Comment (dguy) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: It is so we can call `consumer.seekToBeginning()` many times and then call `consumer.position(..)` many times without incurring a blocking round trip every time. 
The fetch positions are only updated if needed.

**Comment by ijuma** [comment]
Would it make sense to also include https://github.com/apache/kafka/pull/2637?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 6 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Probably worth a comment here since it's not obvious.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 7]
```diff
@@ -244,11 +246,11 @@ public void updateFetchPositions(Set<TopicPartition> partitions) {
                 continue;
 
             if (subscriptions.isOffsetResetNeeded(tp)) {
-                resetOffset(tp);
+                resetOffsets(Collections.singleton(tp));
             } else if (subscriptions.committed(tp) == null) {
                 // there's no committed position, so we need to reset with the default strategy
                 subscriptions.needOffsetReset(tp);
-                resetOffset(tp);
+                resetOffsets(Collections.singleton(tp));
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I guess there's still room to batch the offset lookup here. If it's too much trouble here, maybe we can open a new JIRA so that we don't forget about it?

**Comment by dguy** [comment]
@guozhangwang @hachikuji is this ok to merge yet?

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch! In truth, I thought that TODO would be there forever.

**Comment by guozhangwang** [comment]
LGTM! Thanks.

**Comment by guozhangwang** [comment]
Merged to trunk.
